hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/translators/BlockStreamTransactionTranslator.java

```diff
@@ -88,72 +112,215 @@ public SingleTransactionRecord translate(
      */
     @Override
     public List<SingleTransactionRecord> translateAll(
-            @NonNull final List<SingleTransactionBlockItems> transactions, @NonNull final StateChanges stateChanges) {
-        throw new NotImplementedException();
+            @NonNull final List<SingleTransactionBlockItems> transactions,
+            @NonNull final List<StateChanges> stateChanges) {
+        return transactions.stream()
+                // Filter out any objects that weren't parsed correctly (if t.txn() is null, that transaction definitely
+                // wasn't parsed correctly)
+                .filter(t -> t.txn() != null)
+                .map(txn -> {
+                    final var consensusTimestamp = txn.result().consensusTimestamp();
+                    final var matchingTimestampChanges = stateChanges.stream()
+                            .filter(stateChange -> Objects.equals(consensusTimestamp, stateChange.consensusTimestamp()))
+                            .toList();
+                    final var matchingChanges =
+                            matchingTimestampChanges.isEmpty() ? null : matchingTimestampChanges.getFirst();
+                    return translate(txn, matchingChanges);
+                })
+                .toList();
     }
 
     private TransactionRecord.Builder parseTransaction(
-            final Transaction txn, final TransactionRecord.Builder recordBuilder) {
-        return recordBuilder
-                .transactionID(txn.body().transactionID())
-                .memo(txn.body().memo());
+            final Transaction txn,
+            final TransactionRecord.Builder recordBuilder,
+            final TransactionReceipt.Builder receiptBuilder)
+            throws NoSuchAlgorithmException {
+        TransactionID transactionID;
+        if (txn.body() != null) {
+            transactionID = pbjToProto(
+                    txn.body().transactionID(), com.hedera.hapi.node.base.TransactionID.class, TransactionID.class);
+            recordBuilder.setTransactionID(transactionID).setMemo(txn.body().memo());
+        } else {
+            final var parts = TransactionParts.from(fromPbj(txn));
+            transactionID = parts.body().getTransactionID();
+            recordBuilder.setTransactionID(transactionID);
+
+            String memo = parts.body().getMemo();
+            if (memo == null || memo.isEmpty()) {
+                final var memoBytes = parts.body().getMemoBytes();
+                if (memoBytes != null && !memoBytes.isEmpty()) {
+                    memo = parts.body().getMemoBytes().toString();
+                }
+            }
+            recordBuilder.setMemo(memo);
+        }
+
+        final var txnBytes = toBytesForHash(txn);
+        final var hash = txnBytes != Bytes.EMPTY ? hashTxn(txnBytes) : Bytes.EMPTY;
+        recordBuilder.setTransactionHash(toByteString(hash));
+
+        receiptBuilder.setAccountID(transactionID.getAccountID());
```

Comment:
I don't think we need to set the `accountId` for every transaction e.g. I don't see it in the records of a `ConsensusSubmitMessage`.
Apart from `CryptoCreate` and hollow account finalization I don't see it used anywhere else in the codebase.

Reply:
My mistake, I was thinking this was the payer ID. Thanks for pointing this out!